server/WFCORE-6750-unstable-api-annotation-scanner.adoc

```diff
@@ -0,0 +1,243 @@
+
+Both the two alternative ideas above would need this conversion to strings, since these `CONSTANT_Utf8_Info` entries contain the names of the classes and members.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFCORE-6750[WFCORE-6750 - Scan for annotations indicating unstable API on deployment]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/WFLY-19152[WFLY-19152 - Add indices of APIs annotated with annotations indicating unstable API]
+
+=== Stability Level
+// Choose the planned stability level for the proposed functionality
+* [ ] Experimental
+
+* [x] Preview
+
+* [ ] Community
+
+* [ ] default
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+* mailto:pferraro@redhat.com[Paul Ferraro] (SME)
+* mailto:ehugonne@redhat.com[Emmanuel Hugonnet] (Outside Perspective)
+
+=== QE Contacts
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+* [x] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+* WildFly
+** the wildfly-galleon-pack and wildfly-ee-feature pack feature packs will scan their classpaths and provide the indices of API elements annotated with `io.smallrye.common.annotation.Experimental` and `org.hibernate.Incubating`.
+* WildFly Core
+** scans the applications when they are deployed, looking for usage of API elements that exist in the index.
+* https://github.com/wildfly/unstable-api-annotation-utils[wildfly/unstable-api-annotation-utils] - contains a library providing:
+** A Maven plugin which can be used by the feature packs to do the scanning
+** A highly optimised bytecode scanner and lookup mechanism to find usage in user applications of API elements that appear in the indices created by the feature packs
+
+
+=== Other Interested Projects
+
+Feature packs external to WildFly, as mentioned in the 'Future Work' section.
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+* Note this is at preview level only!
+* There is a maven plugin that can be used in the feature pack builds.
+** It will scan for annotations indicating unstable API, and add the following to the `content/` directory of the  `{indexModule}` module (the `content/` directory is defined as a resource root:
+*** a file containing the indices for annotations concerning the feature pack.
+*** The name of the file will be added to a file called `index.txt` in the `content/` folder of the `{indexModule}` module. This `index.txt` is provided as the main lookup mechanism for the files mentioned in the above point.
```

Comment:
Yuck. I don't like the idea of FPs coordinating changes to a single file.

Perhaps use ModuleClassLoader.iterateResources() or some other API in MCL to remove the need for this index.txt.

Reply:
Kabir implemented this idea.